core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java

```diff
@@ -141,25 +137,15 @@ private static void putReportedMetrics(IndexedSet<Metric> metricSet, String inst
       // If a metric is COUNTER, the value sent via RPC should be the incremental value; i.e.
       // the amount the value has changed since the last RPC. The master should equivalently
       // increment its value based on the received metric rather than replacing it.
-      if (metric.getMetricType() == MetricType.COUNTER) {
-        // FULL_NAME_INDEX is a unique index, so getFirstByField will return the same results as
-        // getByField
-        Metric oldMetric = metricSet.getFirstByField(FULL_NAME_INDEX, metric.getFullMetricName());
-        double oldVal = oldMetric == null ? 0.0 : oldMetric.getValue();
-        Metric newMetric = new Metric(metric.getInstanceType(), metric.getHostname(),
-            metric.getMetricType(), metric.getName(), oldVal + metric.getValue());
-        for (Map.Entry<String, String> tag : metric.getTags().entrySet()) {
-          newMetric.addTag(tag.getKey(), tag.getValue());
-        }
-        metricSet.removeByField(FULL_NAME_INDEX, metric.getFullMetricName());
-        newMetrics.add(newMetric);
+      Metric oldMetric = metricSet.getFirstByField(FULL_NAME_INDEX, metric.getFullMetricName());
+      if (oldMetric == null) {
+        metricSet.add(metric);
+      } else if (metric.getMetricType() == MetricType.COUNTER) {
+        oldMetric.addValue(metric.getValue());
       } else {
-        metricSet.removeByField(FULL_NAME_INDEX, metric.getFullMetricName());
-        newMetrics.add(metric);
+        oldMetric.setValue(metric.getValue());
       }
     }
-    metricSet.removeByField(ID_INDEX, instanceId);
```

Comment:
Is there a case where a field will actually be removed? Looking at the javadoc for this method, it seems like there was a case where fields would be removed?

Reply:
This related to the MetricsMasterTest.testAggregater. For a worker or client A, if it reports metrics1 and metrics2 in previous heartbeats, but only reports metrics1 in the current heartbeat, the metric recording of the metric2 of this instance A will be deleted.